Add Cask tutorials (#3056)
scala/docs.scala-lang: _overviews/toolkit/web-server-static.md

---

> title: How to serve a static file?

Author: I would like to preserve the "How to [do X]" title convention, used in all the other tutorials. I imagine that's what a person in a need for guidance would search for.

```
{% endtab %}
{% endtabs %}

The `@cask.staticFiles` annotation specifies at which path the webpage will be available. The endpoint function returns
the location in which the file can be found.

{% tabs web-server-static-2 class=tabs-scala-version %}
{% tab 'Scala 2' %}
```scala
object Example extends cask.MainRoutes {
  @cask.staticFiles("/static")
  def staticEndpoint(): String = "src/main/resources" // or "resources" if not using SBT

  initialize()
}
```
{% endtab %}
{% tab 'Scala 3' %}
```scala
object Example extends cask.MainRoutes:
  @cask.staticFiles("/static")
  def staticEndpoint(): String = "src/main/resources" // or "resources" if not using SBT

  initialize()
```
{% endtab %}
{% endtabs %}

In the example above, `@cask.staticFiles` instructs the server to look for files accessed at the `/static` path in the 
`src/main/resources` directory. Cask will match any subpath coming after `/static` and append it to the directory path.
If you access the `/static/hello.html` file, it will serve the file available at `src/main/resources/hello.html`.
The directory path can be any path available to the server, relative or not. If the requested file cannot be found in the

> specified location, a 404 response with an error message will be returned instead.

Reviewer: ```suggestion
specified location, the server will return a 404 response with an error message.
```

> The `Example` object inherits from the `cask.MainRoutes` class, providing the main function that starts the server. The `initialize()`

Reviewer: ```suggestion
The `Example` object inherits from the `cask.MainRoutes` class. It provides the main function that starts the server. The `initialize()`
```

method call initializes the server routes, i.e., the association between URL paths and the code that handles them.

### Using the resources directory

The `@cask.staticResources` annotation works in the same way as the `@cask.staticFiles` used above, with the difference that
the path returned by the endpoint method describes the location of files _inside_ the resources directory. Since the
previous example conveniently used the resources directory, it can be simplified with `@cask.staticResources`.

{% tabs web-server-static-3 class=tabs-scala-version %}
{% tab 'Scala 2' %}
```scala
object Example extends cask.MainRoutes {
  @cask.staticResources("/static")
  def staticEndpoint(): String = "."

  initialize()
}
```
{% endtab %}
{% tab 'Scala 3' %}
```scala
object Example extends cask.MainRoutes:
  @cask.staticResources("/static")
  def staticEndpoint(): String = "."

  initialize()
```
{% endtab %}
{% endtabs %}

In the endpoint method, the location is set to `"."`, telling the server that the files are available directly in the

> resources directory. In general, you can use any nested location within the resources directory, for instance you could opt

Reviewer: ```suggestion
resources directory. In general, you can use any nested location within the resources directory. For instance, you could opt
```
